### ISSUE TITLE
A note of thank you and concern

### ISSUE THREAD

**Issue by theodor38** [ludwig-ai/ludwig#603]
I will start with saying 'sorry' because this is not a feature request but a thank you for what has been developed into Ludwig and a concern about its slow updates(commits) over the last month.

There are scantly any day which I don't refresh Ludwig Github to see if there is any new updates. Hence me noticing the commits had become less and less frequent.

When I found out about Ludwig I was ecstatic. This is exactly what is needed and already being done in other respects with stable baselines for RL and Optuna bayesian optimization libraries.  

Please continue developing this treasure. I am not being able to contribute as such a thing I feel is beyond me at this point technically. Having said that I have no doubt that Ludwig will capture more attention as it progresses. 

Thank you again and best of luck with continued development!

(Feel free to delete this as it is really not about a feature request)

**Comment by ifokeev** [comment]
I think it's a good idea for donations. I may contribute too, but have no free time

**Comment by theodor38** [comment]
If there was a way to donate, I would for sure.

**Comment by ifokeev** [comment]
@w4nderlust so can you open github donations for this repo?

**Comment by msaisumanth** [comment]
@theodor38 Thanks for raising this. Rest assured, we are actively working on Ludwig. We have been busy with conferences and internal deadlines.

**Comment by pingsutw** [comment]
@msaisumanth I want to contribute this repo, do you have any roadmap about this repo.

**Comment by msaisumanth** [comment]
@pingsutw Yes, we welcome contributions from everyone. But we currently don't have a public agenda. I will provide an update on it after a team discussion.

In the meantime, please take a look at the [open issues](https://github.com/uber/ludwig/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22help+wanted%22), [enhancements](https://github.com/uber/ludwig/issues?q=is%3Aopen+is%3Aissue+label%3Aenhancement) and see if anything interests you. Also, you're welcome to propose new ideas too!

**Comment by pingsutw** [comment]
@msaisumanth Thanks for the reply, I will find some work in open issues first. I'm glad to heard any update about this repo after your discussion

**Comment by w4nderlust** [comment]
Dear @theodor38 ,

thank you very much for this post and for your kind words. It may be difficult to perceive, but working on an open source project like this can be overwhelming, and most of the communications we receive are about errors or bugs to be solved, words of appreciation like yours are rare, but very much appreciated and help substantially in keeping us motivated!

Be sure that Ludwig is not going anywhere and we are really much commited to work on it and improve it. The recent slow down in terms of commits is due, as @msaisumanth already explained, to an alignment of deadlines and travels (I've personally attended more than 6 conferences in 4 different conferences in the second half of this year). At the same time, we are working a lot behind the scenes on a major change in the preprocessing which will make it possible to train Ludwig models on many more different data formats, and in particular to train on datasets beyond memory / disk size. As this change is not a small incremental change, but a quite substantial change, it doesn't make a lot of sense to release pieces of it, we prefer to release the finalized change, so the maount of work that is going into it is not entirely visible to the community.

Finally, we have big plans for 2020, this change in the preprocessing is the first of a number of improvements that you will see, which will include porting to TensorFlow 2 which will enable much more pre-trained model and flexibility in adding new models. we are also going to support hyperparameter optimization, something we are also working on behind the scenes. Any help from the community will be more than welcome, and it will probably have to start from us to begin with, with sharing our plans to open up development even more.

I'd like to thank @ifokeev too for proposing donations. We deeply appreciate it, but there is no need for it as it would not change our development process. We will reorganize our work in early 2020 to be able to spend more time on Ludwig than we have been able to the last couple months of 2019.

Thank you again for your continued support.
- Piero

**Comment by theodor38** [comment]
@w4nderlust @msaisumanth 

Thank you much for the update. I wish you a great new year and best of luck with continued development with Ludwig.

In time I will try to contribute as well as I get better myself technically speaking. 

Kind regards 
Memo

**Comment by ifokeev** [comment]
@w4nderlust by the way, what is the ETA to say thank you for `TF 2`? ;)

**Comment by w4nderlust** [comment]
It's the second big thing we will be focusing on after we are done with preprocessing. Hopefully we will be done with preprocessing by the end of January and work on TF2 in February. Can't be 100% sure we'll finish by the end of February, bu that's what we are shooting for.

**Comment by theodor38** [comment]
I have a quick question. 

Does Ludwig team have any plans to integrate any multimodel learning?

Is multimodel learning a practical reality as of yet? I know Google had published the "one model to learn them all" paper in 2017 but has it been integrated into TensoFlow and what not?

I had been really curious about multimodel learning. What are your opinions on it?

Memo

**Comment by theodor38** [comment]
Thank you for the Reply!

Once with Ludwig I tried multimodality such that I created a dataset of  rows of two numbers from 1 to 1000 and an output as their addition(simple arithmetic) and whether the result would be odd or even. So in this case the multimodality was classification and regression at the same time. 

However, I felt like there were two separate networks being trained and I was receiving two sequential answers one for continuous number and other for classification. 

This lead me to believe that it wasn't a single network doing the regression and classification at the same time but two separate ones. 

Maybe I did it no the right way or misinterpreted the "verbose" visual cues. I should have tried to save the models to see if there were more than one. 

In any case, it is my uderstanding that with multimodality all different types of inputs are given to a single network and that single network predicting the results. 

Like I said this could all be due to my inexperience. Now that I think about it, there wasn't any multimodality in my input space but I was expecting two different answers from the same network.

**Comment by w4nderlust** [comment]
If you specify multiple inputs and multiple outputs Ludwig has a separate encoder for each input and a separate decoder for each output and has a central piece called combiner that combines information from each input encoder and provides it to all output decoders. So it is one single model, some pieces of it are specific to the each input and outputs and some pieces of it are in common. How much of the model is in common and how much of it is separate depends on the hyperparameters of the different components. I should probably make this more clear in the documentation, but if you look both at the [blogpost](https://eng.uber.com/introducing-ludwig/) or the [paper](https://w4nderlu.st/publications/ludwig-a-type-based-declarative-deep-learning-toolbox/ludwig-a-type-based-declarative-deep-learning-toolbox.pdf) or the [slides](https://w4nderlu.st/projects/ludwig/ludwig_presentation.pdf) it may become more clear :)

**Comment by theodor38** [comment]
Piero, 

Thank you very much for your explanation. It makes a lot of sense

I will go through the links you have sent also. 

All the best

**Comment by w4nderlust** [comment]
I spent some time coming up with a project Kanban here in the project tab on GitHub: https://github.com/uber/ludwig/projects/1
Hopefully this will give more visibility on what we are working on, on priorities and encourage contribution by collaborating in the open.

**Comment by theodor38** [comment]
In these times in which we don't hear too many good news, I wanted to congratulate the Ludwig team and all who had contributed on the new version!.

Be safe and thank you for all the hard work!

**Comment by w4nderlust** [comment]
This is heartwarming @theodor38 , thank you so much!

There's a lot work that is also being put in the next version (you can look at the various TF2 related PRs) coming fomr several contributors, in particular @jimthompson5802 . Brace your self, many things will be changed and improved, and it will be the best version of Ludwig yet!
